### PR TITLE
Update project deployment tooling, add docker-compose development environment

### LINKED ISSUE #223: Make project deployment tooling consistent
Convert project structure and deployment tooling so that it aligns with the following projects:

- https://github.com/raster-foundry/raster-foundry-docs
- https://github.com/azavea/cicero-live
- https://github.com/azavea/fund-equity-map

This consists of overhauling:

- Vagrant setup
- Deployment tooling
- AWS resource management with Terraform
- Removal of `s3_website`

### PR DESCRIPTION
**Pull request by tnation14**
## Overview

This PR brings this project into alignment with our standard setup by updating the deployment tooling to use Terraform and AWS instead of the `s3_website` gem, and replacing the Vagrant development environment with a docker-compose based setup. 

### Notable Changes
- [x] Repace `npm` package manager with `yarn`
- [x] Replace Vagrant setup with docker-compose
- [x] Replace `s3_website` deployment with Terraform/AWS CLI 
- [x] Add docker-based deployment tooling

### Demo

~Optional. Screenshots, `curl` examples, etc.~

### Notes
- I couldn't verify that `package-lock.json` is exactly the same as yarn.lock. However, I created yarn.lock using the following steps:
```bash
$ npm install # create node_modules from package.json
$ yarn import # create yarn.lock from the contents of the node_modules folder
```

- Initially, dockerized Karma tests were failing in CI (Ubuntu 16.04) with the following error:
```bash
ChromeHeadless stderr: Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Operation not permitted
```
I was able to get things working by creating a Karma custom launcher that passes `--no-sandbox` to Chomium. The custom launcher overrode the default launcher flags, so I had to include `--disable-gpu`, `--headless`, and `--remote-debugging-port` as well. 

## Testing Instructions

 * Rebuild the `app` VM, ensure the process completes successfully.
 * Run `script/server`, make sure the site is visible at http://localhost:4200
 * See [Jenkins output](http://urbanappsci.internal.azavea.com/job/azavea/job/climate-change-lab/job/feature%252Ftnation%252Fupdate-deployment-tooling/17/console)
 * See https://lab.staging.climate.azavea.com

Closes #223

### CODE REVIEW

**Comment by tnation14** [comment]
@CloudNiner could you weigh in on the switch to `yarn` and my modifications to the Karma configuration? Just want to be sure that it's not going to break anything else.

File: deployment/terraform/cdn.tf [hunk 1]
```diff
@@ -0,0 +1,75 @@
+resource "aws_cloudfront_distribution" "cdn" {
```
Comment (tnation14) [inline deployment/terraform/cdn.tf]: Reminder to tear down the old infrastructure once this PR is merged.

**Comment by CloudNiner** [comment]
Is there any significant advantage to `yarn` now that npm has the `package-lock.json` file? 

I'll defer to @flibbertigibbet regarding the karma conf, she was just updating dependencies in #228 so it's probably worth rebasing this against develop once that's merged to ensure you've got the most up to date packages and the testing config here still works.

**Comment by tnation14** [comment]
The main reasons for using `yarn` here are that it alleviates the `node-sass` compilation issue that was being fixed in [ansible/climate-change-lab.app](https://github.com/azavea/climate-change-lab/blob/develop/ansible/roles/climate-change-lab.app/tasks/main.yml#L13-L16), and it's already installed in the `node` Docker image. NPM's installed as well, but it's an older version so we'd have to update it in the Dockerfile.

Re: Karma, I'll try testing again after #228 is merged. Thanks!

**Comment by CloudNiner** [comment]
👍  to yarn then.

**Comment by CloudNiner** [comment]
I built the VM and then ran `./scripts/server`. The docker containers were pulled/built successfully, but then I got this error:
```
vagrant@app:/vagrant$ ./scripts/server
yarn run v0.24.4
warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
$ ng serve
sh: 1: ng: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

It occurs regardless of whether I use the `--vm` option.

**Comment by tnation14** [comment]
Did you build the VM using `scripts/setup`, or `vagrant up`?

**Comment by CloudNiner** [comment]
I just rebuilt after pulling down the latest commit via `./scripts/setup`. Then ran `./scripts/server` in the VM. The webpack server started.

However, I can't hit localhost:4200 from my browser. Unsure if related, but I'm still seeing this yarn warning:
```
warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
```

**Comment by tnation14** [comment]
Did you run `scripts/server --vm`, or just `scripts/server`?

File: deployment/terraform/storage.tf [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#
+# S3 resources
+#
+data "template_file" "read_only_bucket_policy" {
```
Comment (hectcastro) [inline deployment/terraform/storage.tf]: I believe that we have parts of this modularized in https://github.com/azavea/terraform-aws-s3-origin.

**Comment by hectcastro** [comment]
GitHub isn't letting me leave comments on these lines:

screen shot 2017-09-01 at 10 35 08

**Comment by CloudNiner** [comment]
@tnation14 sorry if you were waiting on me for this, I was able to hit the application after reprovisioning and doing `./scripts/server --vm`. Although I still see the "no internet" error that you're seeing.

**Comment by hectcastro** [comment]
I'm not super worried about the "no internet" error. We see it in all of the other Yarn + Docker projects, but it doesn't appear to have any significant negative effects.

See: https://github.com/yarnpkg/yarn/issues/749

**Pull request merged**